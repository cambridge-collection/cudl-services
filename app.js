//Config
config = require('../config/base.js');
users = require('../config/users.js');

//Modules
var express = require('express');
var path = require('path');
var favicon = require('static-favicon');
var logger = require('morgan');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');
var mysql = require('mysql');
passport = require('passport')
var strategy = require('passport-localapikey').Strategy;
var fs = require('fs-extra');

//cache directories
fs.ensureDir(config.cacheDir);
fs.ensureDir(config.cacheDir+'/transcriptions');
fs.ensureDir(config.cacheDir+'/translations');

//Routes
var routes = require('./routes/index.js');
var metadata = require('./routes/metadata.js');
var transcription = require('./routes/transcription.js');
var translation = require('./routes/translation.js');
var membership = require('./routes/membership.js');
var embedded = require('./routes/embedded.js');
var iiif = require('./routes/iiif.js');
var app = express();

//MySQL Connection
connection = mysql.createPool({
	connectionLimit	: 10,
        host     : config.mysqlHost,
        user     : config.mysqlUser,
        password : config.mysqlPass,
	database : config.mysqlData,
});


function findByApiKey(apikey, fn) {
    if (apikey in users) {
      return fn(null, users[apikey]);
    }
  return fn(null, null);
}


passport.use(new strategy(
  function(apikey, done) {
    process.nextTick(function () {
      findByApiKey(apikey, function(err, user) {
        if (err) { return done(err); }
        if (!user) { return done(null, false, { message: 'Unknown apikey : ' + apikey }); }
        return done(null, user);
      })
    });
  }
));


// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'jade');

app.use(favicon(path.join(__dirname, '/public/images/brand/favicon.ico')));
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded());
app.use(cookieParser());
app.use(passport.initialize());
app.use(express.static(path.join(__dirname, 'public')));

//app.use('/', routes);
app.use('/v1/metadata', metadata);
app.use('/v1/transcription',transcription);
app.use('/v1/translation', translation);
app.use('/v1/rdb/membership', membership);
app.use('/v1/embedded', embedded);
app.use('/v1/iiif', iiif);

/// catch 404 and forward to error handler
app.use(function(req, res, next) {
    var err = new Error('Not Found');
    err.status = 404;
    next(err);
});

/// error handlers

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
    app.use(function(err, req, res, next) {
        res.status(err.status || 500);
        res.render('error', {
            message: err.message,
            error: err
        });
    });
}

// production error handler
// no stacktraces leaked to user
app.use(function(err, req, res, next) {
    res.status(err.status || 500);
    res.render('error', {
        message: err.message,
        error: {}
    });
});

console.log('hmmm');
module.exports = app;
