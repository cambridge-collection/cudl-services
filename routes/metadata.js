var express = require('express');
var fs = require("fs");
var util = require("util");

var config = require('../config/base');
var serviceUtil = require("../util.js");

var CORS_HEADERS = {
	"Access-Control-Allow-Origin": "*"
};

var router = express.Router();

/* GET home page. */
router.get('/', function(req, res) {
    res.status(401).send('Unathorised');
/*  res.render('index', { title: 'Unathorised' });*/
});

router.get('/:format/:id',
	function(req, res) {
		// We always want to allow remote ajax access
		res.set(CORS_HEADERS);

		// The response depends on the Origin header, as we block access to
		// non-embeddable items from non-cudl origins. If we don't set
		// Vary: Origin then a response for a client on CUDL could be used by a
		// cache to service a request from an external site.
		res.set("Vary", "Origin");

		// Ensure our ID doesn't contain slashes, as it's used to build
		// filesystem paths.
		if(/\//.test(req.params.id)) {
			res.status(400).json({
				error: util.format("Bad id: %s", req.params.id)
			});
			return;
		}

		var path;

		if (req.params.format == 'json') {
			path = config.dataDir + '/json/' + req.params.id + '.json';
			loadJsonMetadata(path, function(err, data) {
				if(err) {
					res.status(404).json({
						error: util.format(
							"ID does not exist: %s", req.params.id)
					});
					return;
				}

				// If the request is an external CORS request we'll restrict
				// access to items which are not embeddable. This prevents
				// external sites using CORS request to get at non-embeddable
				// content, while allowing cudl itself to get at it. Note that
				// there's nothing to stop someone setting up a proxy which
				// strips or fakes the origin header.
				if(serviceUtil.isExternalCorsRequest(req) &&
						data.embeddable === false) {
					res.status(403).json({
						error: "This item is only available from " +
								"cudl.lib.cam.ac.uk"
					});
					return;
				}

				res.json(data);
			});
		} else {
			path = config.dataDir+'/data/'+req.params.format+'/'+req.params.id+'/'+req.params.id+'.xml';
 			res.sendfile(path);
  		}
	}
);

function loadJsonMetadata(path, cb) {
	fs.readFile(path, "utf-8", function(err, data) {
		if(err) cb(err);
		try{
			var parsed = JSON.parse(data);
		}
		catch(e) {
			cb(util.format("Error parsing file as JSON: %s. %s", path, e));
			return;
		}
		cb(null, parsed);
	});
}

module.exports = router;
