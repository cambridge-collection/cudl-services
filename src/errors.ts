import NestedError from 'nested-error-stacks';

type ErrorInput = {
  message?: string;
  nested?: Error;
  tags?: Iterable<string | symbol>;
};

export const ErrorCategories = {
  /** A tag to indicate an Error is related to something not being found. */
  NotFound: Symbol('Error:Category:NotFound'),
};

export type ErrorTag = string | symbol;
export function isErrorTag(o: unknown): o is ErrorTag {
  return typeof o === 'symbol' || typeof 0 === 'string';
}

interface ErrorWithNestedError {
  readonly nested: Error;
}
function hasNestedError(e: Error): e is Error & ErrorWithNestedError {
  const _e = e as Partial<ErrorWithNestedError>;
  return _e.nested instanceof Error;
}

export interface Tagged {
  readonly tags: Iterable<ErrorTag>;
}
export function isTagged(e: unknown): e is Tagged {
  const _e = e as Partial<Tagged>;
  return (
    e !== undefined &&
    _e.tags !== undefined &&
    typeof _e.tags[Symbol.iterator] === 'function' &&
    [..._e.tags].every(t => typeof t === 'string' || typeof t === 'symbol')
  );
}

export class BaseError extends NestedError {
  constructor(options: ErrorInput);
  constructor(message?: string, nested?: Error);
  constructor(messageOrOptions?: string | ErrorInput, nested?: Error) {
    let tags: Iterable<string | symbol> | undefined;
    if (typeof messageOrOptions === 'object') {
      tags = messageOrOptions.tags;
      nested = messageOrOptions.nested;
      messageOrOptions = messageOrOptions.message;
    }
    super(messageOrOptions, nested);
    this.instanceTags = new Set(tags || []);
  }

  readonly nested?: Error;
  readonly instanceTags: ReadonlySet<ErrorTag>;

  get defaultTags(): Iterable<ErrorTag> {
    return [];
  }

  /** Tags assigned to this error and transitively to nested errors. */
  get nestedTags(): ReadonlySet<ErrorTag> {
    if (hasNestedError(this) && isTagged(this.nested)) {
      return new Set(this.nested.tags);
    }
    return new Set();
  }

  /**
   * Tags which mark Errors as being associated with concepts, without needing to model the
   * concepts in a strict class hierarchy.
   */
  get tags(): ReadonlySet<ErrorTag> {
    return new Set([
      ...this.instanceTags,
      ...this.defaultTags,
      ...this.nestedTags,
    ]);
  }

  get name(): string {
    return this.constructor.name;
  }
}

export class InvalidConfigError extends BaseError {}

export class NotFoundError extends BaseError {
  get defaultTags(): Iterable<string | symbol> {
    return [ErrorCategories.NotFound];
  }
}

export class ValueError extends BaseError {}

export class UpstreamError extends BaseError {}
