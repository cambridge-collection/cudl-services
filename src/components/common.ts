import {Component, MiddlewareComponent, SettingsComponent} from '../app';
import path from 'path';
import bodyParser from 'body-parser';
import express from 'express';
import {
  passportApiKeyAuthenticationComponent,
  passportInitialiseComponent,
  Users,
} from './api-key-auth';

const cookieParser = require('cookie-parser');
const favicon = require('serve-favicon');
const logger = require('morgan');

/**
 * Components to register ahead of regular request handling middleware.
 *
 * @param options
 */
export function leadingComponents(options?: {apiKeys?: Users}): Component[] {
  options = options || {};

  const authMiddleware =
    options.apiKeys === undefined
      ? []
      : [
          passportInitialiseComponent,
          passportApiKeyAuthenticationComponent(options.apiKeys),
        ];

  return [
    // Disable parsing of query strings into nested objects
    new SettingsComponent({'query parser': 'simple'}),
    new MiddlewareComponent({
      handler: [
        favicon(
          path.resolve(__dirname, '../../public/images/brand/favicon.ico')
        ),
        logger('dev'),
        bodyParser.json(),
        bodyParser.urlencoded({extended: false}),
        cookieParser(),
        express.static(path.resolve(__dirname, '../../public')),
        stripTrailingSlashes,
      ],
    }),
    ...authMiddleware,
  ];
}

/**
 * Middleware to redirect trailing slashes to same URL without trailing slash
 */
const stripTrailingSlashes: express.Handler = (req, res, next) => {
  if (req.url.substr(-1) === '/' && req.url.length > 1) {
    res.redirect(301, req.url.slice(0, -1));
  } else {
    next();
  }
};
