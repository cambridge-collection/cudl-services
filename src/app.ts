import { XSLTExecutor } from '@lib.cam/xslt-nailgun';
import bodyParser from 'body-parser';
import Debugger from 'debug';
import express, { NextFunction, Request, Response } from 'express';
import passport from 'passport';
import path from 'path';
import { CollectionDAO, PostgresCollectionDAO } from './collections';
import { Config, User, Users } from './config';

import { DAOPool, PostgresDatabasePool } from './db';
import {
  CUDLMetadataRepository,
  DefaultCUDLMetadataRepository,
  LegacyDarwinMetadataRepository,
} from './metadata';
import { BaseResource, ExternalResources, using } from './resources';
import * as darwin from './routes/darwin';
import * as membership from './routes/membership';
import * as metadata from './routes/metadata';
import * as similarity from './routes/similarity';
import * as tags from './routes/tags';
import { PostgresTagsDAO, TagsDAO } from './routes/tags-impl';
import * as transcription from './routes/transcription';
import * as translation from './routes/translation';
import { DefaultXTF, XTF } from './xtf';

const cookieParser = require('cookie-parser');
const favicon = require('serve-favicon');
const logger = require('morgan');
const Strategy = require('passport-accesstoken').Strategy;

const debug = Debugger('cudl-services');

// FIXME: move this to server?
//cache directories
// fs.ensureDirSync(config.cacheDir);
// fs.ensureDirSync(config.cacheDir+'/transcriptions');
// fs.ensureDirSync(config.cacheDir+'/translations');

export interface AppOptions {
  users: Users;
  metadataRepository: CUDLMetadataRepository;
  legacyDarwinMetadataRepository: LegacyDarwinMetadataRepository;
  collectionsDAOPool: DAOPool<CollectionDAO>;
  tagsDAOPool: DAOPool<TagsDAO>;
  darwinXtfUrl: string;
  xtf: XTF;
}

export class App extends BaseResource {
  readonly options: AppOptions;
  readonly expressApp: express.Application;
  protected readonly xsltExecutor: XSLTExecutor;

  constructor(options: AppOptions) {
    super();
    this.options = Object.freeze({ ...options });
    this.xsltExecutor = XSLTExecutor.getInstance();
    this.expressApp = this.createExpressApp();
  }

  static fromConfig(config: Config): ExternalResources<App> {
    const dbPool = PostgresDatabasePool.fromConfig(config);

    return new ExternalResources(
      new App({
        metadataRepository: new DefaultCUDLMetadataRepository(config.dataDir),
        legacyDarwinMetadataRepository: new LegacyDarwinMetadataRepository(
          config.legacyDcpDataDir
        ),
        collectionsDAOPool: PostgresCollectionDAO.createPool(dbPool),
        tagsDAOPool: PostgresTagsDAO.createPool(dbPool),
        users: config.users,
        darwinXtfUrl: config.darwinXTF,
        xtf: new DefaultXTF(config),
      }),
      [dbPool]
    );
  }

  private createExpressApp(): express.Application {
    const app = express();

    passport.use(
      new Strategy((token: string, done: (err: any, user: any) => void) => {
        process.nextTick(() => {
          const user = findByApiKey(this.options.users, token);
          return done(null, user || false);
        });
      })
    );

    // view engine setup
    app.set('views', path.resolve(__dirname, '../views'));
    app.set('view engine', 'pug');

    app.use(
      favicon(path.resolve(__dirname, '../public/images/brand/favicon.ico'))
    );
    app.use(logger('dev'));
    app.use(bodyParser.json());
    app.use(bodyParser.urlencoded({ extended: false }));
    app.use(cookieParser());
    app.use(passport.initialize());
    app.use(express.static(path.resolve(__dirname, '../public')));

    // Middleware to redirect trailing slashes to same URL without trailing slash
    app.use((req, res, next) => {
      if (req.url.substr(-1) === '/' && req.url.length > 1) {
        res.redirect(301, req.url.slice(0, -1));
      } else {
        next();
      }
    });

    app.use(
      '/v1/metadata',
      metadata.getRoutes({
        metadataRepository: this.options.metadataRepository,
      })
    );

    app.use(
      '/v1/rdb/membership',
      membership.getRoutes({
        getItemCollections: async (itemID: string) =>
          using(this.options.collectionsDAOPool.getInstance(), dao =>
            dao.getItemCollections(itemID)
          ),
      })
    );

    app.use('/v1/tags', tags.getRoutes({ daoPool: this.options.tagsDAOPool }));

    app.use(
      '/v1/transcription',
      transcription.getRoutes({
        metadataRepository: this.options.metadataRepository,
        legacyDarwinMetadataRepository: this.options
          .legacyDarwinMetadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/translation',
      translation.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/xtf/similarity',
      similarity.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xtf: this.options.xtf,
      })
    );

    app.use(
      '/v1/darwin',
      passport.authenticate('token', { session: false }),
      darwin.getRoutes({ darwinXtfUrl: this.options.darwinXtfUrl })
    );

    // 404 if no route matched
    app.use((req: Request, res: Response, next: NextFunction) => {
      res.status(404).send('Not Found');
    });

    return app;
  }

  async close(): Promise<void> {
    await Promise.all([super.close(), this.xsltExecutor.close()]);
  }
}

function findByApiKey(users: Users, apiKey: string): User | null {
  if (apiKey in users) {
    return users[apiKey];
  }
  return null;
}
