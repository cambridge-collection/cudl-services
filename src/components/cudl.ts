import {DAOPool} from '../db';
import {Components, MiddlewareComponent} from '../app';
import * as tags from '../routes/tags';
import * as translation from '../routes/translation';
import * as similarity from '../routes/similarity';
import {CollectionDAO} from '../collections';
import {
  CUDLMetadataRepository,
  cudlProvidersForDataStore,
} from '../metadata/cudl';
import {URL} from 'url';
import * as transcription from '../routes/transcription';
import {XSLTExecutor} from '@lib.cam/xslt-nailgun';
import * as metadata from '../routes/metadata';
import {DomainNameMatcher, ExternalCorsRequestMatcher} from '../util';
import {DataStore, ItemJsonMetadataResponseEmitter} from '../metadata';
import * as membership from '../routes/membership';
import {using} from '../resources';

export type TagsOptions = Omit<tags.GetRoutesOptions, 'router'>;
export function tagsComponents(options: TagsOptions): Components {
  return new MiddlewareComponent({
    path: '/v1/tags',
    handler: tags.getRoutes({daoPool: options.daoPool}),
  });
}

export interface TranscriptionOptions {
  metadataRepository: CUDLMetadataRepository;
  teiServiceURL: URL;
  xsltExecutor: XSLTExecutor;
  zacynthiusServiceURL: URL;
}
export function transcriptionComponents(
  options: TranscriptionOptions
): Components {
  return new MiddlewareComponent({
    path: '/v1/transcription',
    handler: transcription.getRoutes({
      metadataRepository: options.metadataRepository,
      teiServiceURL: options.teiServiceURL,
      xsltExecutor: options.xsltExecutor,
      zacynthiusServiceURL: options.zacynthiusServiceURL,
    }),
  });
}

export type TranslationOptions = Omit<translation.GetRoutesOptions, 'router'>;
export function translationComponents(options: TranslationOptions): Components {
  return new MiddlewareComponent({
    path: '/v1/translation',
    handler: translation.getRoutes(options),
  });
}

export interface MetadataOptions {
  cudlDataDataStore: DataStore;
  internalDomainName: string;
}
export function metadataComponents(options: MetadataOptions): Components {
  return new MiddlewareComponent({
    path: '/v1/metadata',
    handler: metadata.getRoutes({
      metadataProviders: new Map(
        Object.entries(cudlProvidersForDataStore(options.cudlDataDataStore))
      ),
      isExternalCorsRequest: ExternalCorsRequestMatcher({
        internalDomains: DomainNameMatcher(options.internalDomainName),
      }),
      metadataEmitters: [ItemJsonMetadataResponseEmitter.instance],
    }),
  });
}

export interface CollectionMembershipOptions {
  collectionsDAOPool: DAOPool<CollectionDAO>;
}
export function collectionMembershipComponents(
  options: CollectionMembershipOptions
): Components {
  return new MiddlewareComponent({
    path: '/v1/rdb/membership',
    handler: membership.getRoutes({
      getItemCollections: async (itemID: string) =>
        using(options.collectionsDAOPool.getInstance(), dao =>
          dao.getItemCollections(itemID)
        ),
    }),
  });
}

export type SimilarityOptions = Omit<similarity.GetRouteOptions, 'router'>;
export function similarityComponents(options: SimilarityOptions): Components {
  return new MiddlewareComponent({
    path: '/v1/xtf/similarity',
    handler: similarity.getRoutes(options),
  });
}
