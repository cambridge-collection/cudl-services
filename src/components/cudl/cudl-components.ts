import {Components, ResourceCleanupComponent} from '../../app';
import {
  cudlProvidersForDataStore,
  MetadataProviderCUDLMetadataRepository,
} from '../../metadata/cudl';
import {PostgresCollectionDAO} from '../../collections';
import {PostgresTagsDAO} from '../../routes/tags-impl';
import {closingOnError} from '../../resources';
import {XSLTExecutor} from '@lib.cam/xslt-nailgun';
import {
  DarwinProxyComponentOptions,
  darwinProxyComponents,
} from '../darwin-correspondence-project';
import {PostgresDatabasePool} from '../../db';
import {DataStore} from '../../metadata';
import {XTF} from '../../xtf';
import {URL} from 'url';
import {
  collectionMembershipComponents,
  metadataComponents,
  similarityComponents,
  tagsComponents,
  transcriptionComponents,
  translationComponents,
} from '../cudl';

const CUDL_HOST = 'cudl.lib.cam.ac.uk';

export interface CudlOptions {
  dbPool: PostgresDatabasePool;
  cudlDataDataStore: DataStore;
  darwin: DarwinProxyComponentOptions;
  internalDomainName?: string;
  teiServiceURL: URL;
  xtf: XTF;
  zacynthiusServiceURL: URL;
  iiifBaseURL?: string;
}

export async function cudlComponents(
  options: CudlOptions
): Promise<Components> {
  const {
    cudlDataDataStore,
    dbPool,
    teiServiceURL,
    xtf,
    zacynthiusServiceURL,
  } = options;
  const internalDomainName = options.internalDomainName || CUDL_HOST;
  const metadataProviders = cudlProvidersForDataStore(
    options.cudlDataDataStore
  );
  const metadataRepository = new MetadataProviderCUDLMetadataRepository(
    metadataProviders
  );
  const collectionsDAOPool = PostgresCollectionDAO.createPool(dbPool);
  const tagsDAOPool = PostgresTagsDAO.createPool(dbPool);

  return await closingOnError(XSLTExecutor.getInstance(), xsltExecutor => [
    tagsComponents({daoPool: tagsDAOPool}),
    transcriptionComponents({
      metadataRepository,
      teiServiceURL,
      xsltExecutor,
      zacynthiusServiceURL,
    }),
    translationComponents({
      teiServiceURL,
      zacynthiusServiceURL: options.zacynthiusServiceURL,
    }),
    metadataComponents({cudlDataDataStore, internalDomainName}),
    collectionMembershipComponents({
      collectionsDAOPool,
    }),
    similarityComponents({xtf, metadataRepository}),
    darwinProxyComponents(options.darwin),
    ResourceCleanupComponent.closing(xsltExecutor),
  ]);
}
