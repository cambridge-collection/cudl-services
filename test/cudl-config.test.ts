import {Application, Component} from '../src/app';
import fs from 'fs';
import glob from 'glob';

import {file, FileResult, withFile} from 'tmp-promise';
import {mocked} from 'ts-jest/utils';
import {promisify} from 'util';
import cudlConfigDefaultExport, {
  ApplicationWithResources,
  CONFIG_FILE_ENVAR,
  CONFIG_JSON_ENVAR,
  configPathsFromEnvar,
  ConfigSource,
  CUDLConfig,
  CUDLConfigData,
  getCudlDataDataStore,
  loadConfigFile,
  loadConfigFromEnvar,
  mergeConfigs,
  parseConfigURLValue,
  PartialCUDLConfigData,
  splitEnvarPaths,
} from '../src/cudl-config';
import {InvalidConfigError} from '../src/errors';
import {ExternalResources, Resource} from '../src/resources';
import express from 'express';
import {URL} from 'url';
import {FilesystemDataStore} from '../src/metadata/filesystem';
import {S3DataStore} from '../src/metadata/s3';
import {cudlComponents} from '../src/components/cudl/cudl-components';
import {MockComponent} from './mocking/local';
import {leadingComponents} from '../src/components/common';

jest.mock('glob', () => {
  return jest.fn(() => {
    throw new Error('No mock implementation specified');
  });
});
jest.mock('pg');
jest.mock('../src/components/cudl/cudl-components');
jest.mock('../src/components/common');

import ProcessEnv = NodeJS.ProcessEnv;

const EXAMPLE_CONFIG: CUDLConfigData = {
  darwinXTF: 'example',
  dataLocation: 'example',
  postDatabase: 'example',
  postHost: 'example',
  postPass: 'example',
  postPort: 1234,
  postUser: 'example',
  teiServiceURL: 'http://tei.example.com/',
  users: {},
  xtfBase: 'example',
  xtfIndexPath: 'example',
  zacynthiusServiceURL: 'http://zac.example.com/',
};

describe('config', () => {
  describe('mergeConfigs()', () => {
    test('merges 1 config', () => {
      const config = mergeConfigs([{source: 'foo', config: EXAMPLE_CONFIG}]);
      expect(config).toEqual(EXAMPLE_CONFIG);
    });

    test('merges multiple configs', () => {
      const configA: PartialCUDLConfigData = {
        ...EXAMPLE_CONFIG,
      };
      delete configA['darwinXTF'];
      const sources: ConfigSource[] = [
        {source: 'a', config: configA},
        {source: 'b', config: {darwinXTF: 'example'}},
      ];

      expect(mergeConfigs(sources)).toEqual(EXAMPLE_CONFIG);
    });

    test('fails if result is not a complete config', () => {
      const sources: ConfigSource[] = [
        {
          source: 'foo',
          config: {
            darwinXTF: 'xxx',
          },
        },
        {
          source: 'bar',
          config: {
            postDatabase: 'xxx',
          },
        },
      ];
      expect(() => mergeConfigs(sources)).toThrowError(InvalidConfigError);
      expect(() => mergeConfigs(sources)).toThrowError(
        /^Failed to load config: Result of merging foo, bar was not a valid config: .+/
      );
    });
  });

  describe('loadConfigFile()', () => {
    let tmpFile: FileResult;
    beforeEach(async () => {
      tmpFile = await file();
    });

    afterEach(() => tmpFile.cleanup());

    test('fails when filePath cannot be read', async () => {
      await expect(loadConfigFile('/does/not/exist')).rejects.toThrow(
        "Failed to load config from file '/does/not/exist': Error: ENOENT: no such file or directory, open '/does/not/exist'"
      );
    });

    test('fails when file contains invalid JSON5', async () => {
      await promisify(fs.writeFile)(tmpFile.fd, '{');
      await expect(loadConfigFile(tmpFile.path)).rejects.toThrow(
        `Failed to load config from file '${tmpFile.path}': SyntaxError: JSON5: invalid end of input at 1:2`
      );
    });

    test.each<[string, PartialCUDLConfigData]>([
      ["/* Comment */ {dataLocation: 'abcd'}", {dataLocation: 'abcd'}],
      [JSON.stringify(EXAMPLE_CONFIG), EXAMPLE_CONFIG],
    ])('returns partial config', async (content, expected) => {
      await promisify(fs.writeFile)(tmpFile.fd, content);
      expect(await loadConfigFile(tmpFile.path)).toEqual(expected);
    });
  });

  test('splitEnvarPaths()', () => {
    expect(splitEnvarPaths(undefined)).toEqual([]);
    expect(splitEnvarPaths('')).toEqual([]);
    expect(splitEnvarPaths(':')).toEqual([]);
    expect(splitEnvarPaths('::')).toEqual([]);
    expect(splitEnvarPaths('foo:')).toEqual(['foo']);
    expect(splitEnvarPaths('foo::')).toEqual(['foo']);
    expect(splitEnvarPaths(':foo')).toEqual(['foo']);
    expect(splitEnvarPaths('a:bc:d')).toEqual(['a', 'bc', 'd']);
  });

  describe('loadConfigFromEnvar()', () => {
    let env: ProcessEnv;
    beforeEach(() => {
      mocked(glob as SimpleGlobSignature).mockClear();
      env = process.env;
      process.env = {};
    });
    afterEach(() => {
      process.env = env;
    });

    type SimpleGlobSignature = (
      pattern: string,
      cb: (err: Error | null, matches: string[]) => void
    ) => void;

    describe('configPathsFromEnvar()', () => {
      test.each<[string[], {}, boolean]>([
        [
          [
            '/etc/cudl-services/config.json?(5)',
            '/etc/cudl-services/conf.d/*.json?(5)',
          ],
          {},
          true,
        ],
        [['/foo'], {[CONFIG_FILE_ENVAR]: '/foo'}, false],
        [['/foo', '/bar/*'], {[CONFIG_FILE_ENVAR]: '/foo:/bar/*'}, false],
      ])(
        'Expands glob patterns %s when environment contains %s',
        async (patterns, env, defaultUsed) => {
          Object.assign(process.env, env);
          mocked(glob as SimpleGlobSignature).mockImplementation((_, cb) =>
            cb(null, ['a', 'b'])
          );

          const result = await configPathsFromEnvar();
          expect(mocked(glob).mock.calls.map(args => args[0])).toEqual(
            patterns
          );
          expect(result.defaultPatternUsed).toBe(defaultUsed);
          expect(result.paths).toEqual(
            new Array(patterns.length)
              .fill(['a', 'b'])
              .reduce((a, b) => a.concat(b))
          );
          expect(result.patterns).toEqual(patterns);
        }
      );
    });

    test('reports failure due to unreadable config module', async () => {
      mocked(glob).mockImplementationOnce(jest.requireActual('glob'));

      await withFile(async tmpFile => {
        process.env[CONFIG_FILE_ENVAR] = tmpFile.path;
        await promisify(fs.chmod)(tmpFile.path, 0o000);
        await expect(loadConfigFromEnvar()).rejects.toThrow(
          `Failed to load config from file '${tmpFile.path}': Error: EACCES: permission denied`
        );
      });
    });

    test('loads config from file identified by envar', async () => {
      mocked(glob).mockImplementation(
        jest.requireActual('glob') as typeof glob
      );

      await withFile(async configA => {
        await withFile(async configB => {
          await promisify(fs.writeFile)(
            configA.fd,
            JSON.stringify(EXAMPLE_CONFIG)
          );
          const configBContent: PartialCUDLConfigData = {
            postPort: 42,
            users: {foo: {email: 'foo@example.com'}},
          };
          await promisify(fs.writeFile)(
            configB.fd,
            JSON.stringify(configBContent)
          );

          const configCContent: PartialCUDLConfigData = {
            users: {foo: {username: 'foo1'}},
            dataLocation: '/other',
          };

          process.env[CONFIG_FILE_ENVAR] = `${configA.path}:${configB.path}`;
          process.env[CONFIG_JSON_ENVAR] = JSON.stringify(configCContent);

          await expect(loadConfigFromEnvar()).resolves.toEqual({
            ...EXAMPLE_CONFIG,
            ...{
              postPort: 42,
              dataLocation: '/other',
              users: {foo: {email: 'foo@example.com', username: 'foo1'}},
            },
          });
        });
      });
    });

    test('loads config envar JSON without config files', async () => {
      mocked(glob as SimpleGlobSignature).mockImplementation((pattern, cb) =>
        cb(null, [])
      );
      process.env[CONFIG_JSON_ENVAR] = JSON.stringify(EXAMPLE_CONFIG);
      await expect(loadConfigFromEnvar()).resolves.toEqual(EXAMPLE_CONFIG);
    });
  });
});

describe('ApplicationWithResources', () => {
  let app: Application;
  let expressApp: express.Express;
  let res1: Resource;
  let res2: Resource;

  beforeEach(() => {
    expressApp = express();
    app = {
      close: jest.fn(),
      expressApp,
    };
    res1 = new ExternalResources(41, []);
    res2 = new ExternalResources(42, []);
    jest.spyOn(res1, 'close');
    jest.spyOn(res2, 'close');
  });

  test("expressApp is wrapped Application's expressApp", () => {
    const awr = ApplicationWithResources.from(app, res1, res2);
    expect(awr.expressApp).toBe(expressApp);
  });

  test('closes Application on close()', async () => {
    const awr = ApplicationWithResources.from(app, res1, res2);
    expect(app.close).not.toHaveBeenCalled();
    expect(res1.close).not.toHaveBeenCalled();
    expect(res2.close).not.toHaveBeenCalled();

    await awr.close();
    expect(app.close).toHaveBeenCalled();
    expect(res1.close).toHaveBeenCalled();
    expect(res2.close).toHaveBeenCalled();
  });
});

describe('parseConfigURLValue', () => {
  test('returns URL instance for valid URL string', () => {
    expect(parseConfigURLValue('http://example.com/', 'example')).toEqual(
      new URL('http://example.com/')
    );
  });

  test('throws error referencing the property name when value is not a valid URL', () => {
    expect(() =>
      parseConfigURLValue('/not/valid', 'invalidProp')
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('getCudlDataDataStore', () => {
  test('returns FilesystemDataStore for non-URL path', () => {
    const dataStore = getCudlDataDataStore({
      dataLocation: '/example dir/path',
    }) as FilesystemDataStore;
    expect(dataStore).toBeInstanceOf(FilesystemDataStore);
    expect(dataStore.rootPath).toEqual('/example dir/path');
  });

  test('returns FilesystemDataStore for file: URL', () => {
    const dataStore = getCudlDataDataStore({
      dataLocation: 'file:///example%20dir/path',
    }) as FilesystemDataStore;
    expect(dataStore).toBeInstanceOf(FilesystemDataStore);
    expect(dataStore.rootPath).toEqual('/example dir/path');
  });

  test('returns S3DataStore for s3: URL', () => {
    const dataStore = getCudlDataDataStore({
      dataLocation: 's3://bucketname/key/prefix/',
    }) as S3DataStore;
    expect(dataStore).toBeInstanceOf(S3DataStore);
    expect(dataStore.options.bucket).toEqual('bucketname');
    expect(dataStore.options.keyPrefix).toEqual('key/prefix/');
  });

  test('throws on unsupported URL', () => {
    expect(() =>
      getCudlDataDataStore({dataLocation: 'foo://abc'})
    ).toThrowErrorMatchingSnapshot();
  });
});

test('module exports an instance of CUDLConfig as the default export', () => {
  expect(cudlConfigDefaultExport).toBeInstanceOf(CUDLConfig);
});

describe('CUDLConfig', () => {
  describe('createApplicationFromConfigData', () => {
    const config: CUDLConfigData = {
      darwinXTF: 'http://darwin.example.com/',
      dataLocation: '/example-data-location',
      postDatabase: 'example-pg-db',
      postHost: 'example-pg-host',
      postPass: 'example-pg-pass',
      postPort: 1234,
      postUser: 'example-pg-user',
      teiServiceURL: 'http://tei.example.com/',
      users: {
        secret: {username: 'example-user', email: 'example@example.com'},
      },
      xtfBase: 'http://xtf.example.com/',
      xtfIndexPath: '/example-xtf-index',
      zacynthiusServiceURL: 'http://zac.example.com/',
    };

    let mockCudlComponents: Component;
    let mockLeadingComponent: Component;

    beforeEach(() => {
      mockCudlComponents = new MockComponent();
      mocked(cudlComponents).mockResolvedValueOnce(mockCudlComponents);
      mockLeadingComponent = new MockComponent();
      mocked(leadingComponents).mockReturnValueOnce([mockLeadingComponent]);
    });

    test('includes cudlComponents in created app', async () => {
      const app = await new CUDLConfig().createApplicationFromConfigData(
        config
      );
      expect(app.components).toContain(mockCudlComponents);
    });

    test('includes common.leadingComponents in created app', async () => {
      const app = await new CUDLConfig().createApplicationFromConfigData(
        config
      );
      expect(app.components).toContain(mockLeadingComponent);
    });
  });
});
