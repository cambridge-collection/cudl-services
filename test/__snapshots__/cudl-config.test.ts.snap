// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCudlDataDataStore throws on unsupported URL 1`] = `"dataLocation URL's protocol is not supported: foo://abc"`;

exports[`parseConfigURLValue throws error referencing the property name when value is not a valid URL 1`] = `"invalidProp is not a valid URL: \\"/not/valid\\". TypeError [ERR_INVALID_URL]: Invalid URL: /not/valid"`;
